src/main/java/org/jitsi/srtp/crypto/OpenSslAesCipherSpi.java

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright @ 2016 - present 8x8, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.srtp.crypto;
+
+import java.security.*;
+import java.security.spec.*;
+import java.util.*;
+import javax.crypto.*;
+import javax.crypto.spec.*;
+
+/**
+ * AES-CTR implementation using OpenSSL via JNI.
+ */
+public class OpenSslAesCipherSpi
+    extends CipherSpi
+{
+    private static final int BLKLEN = 16;
+
+    private Key key;
+
+    private static native long AES_CTR_CTX_create();
+
+    private static native void AES_CTR_CTX_destroy(long ctx);
+
+    private static native boolean AES_CTR_CTX_init(long ctx, byte[] key);
+
+    private static native boolean AES_CTR_CTX_process(long ctx, byte[] iv,
+        byte[] inOut, int offset, int len);
+
+    /**
+     * the OpenSSL AES_CTR context
+     */
+    private long ctx;
+
+    private byte[] iv;
+
+    private AlgorithmParameters parameters;
+
+    public OpenSslAesCipherSpi()
+    {
+        if (!JitsiOpenSslProvider.isLoaded())
+        {
+            throw new RuntimeException("OpenSSL wrapper not loaded");
+        }
+
+        ctx = AES_CTR_CTX_create();
+        if (ctx == 0)
+        {
+            throw new RuntimeException("CIPHER_CTX_create");
+        }
+    }
+
+    @Override
+    protected void engineSetMode(String mode) throws NoSuchAlgorithmException
+    {
+        if (!"ctr".equalsIgnoreCase(mode))
+        {
+            throw new NoSuchAlgorithmException("Only CTR mode is supported");
+        }
+    }
+
+    @Override
+    protected void engineSetPadding(String padding)
+        throws NoSuchPaddingException
+    {
+        if (!"nopadding".equalsIgnoreCase(padding))
+        {
+            throw new NoSuchPaddingException("No padding support");
+        }
+    }
+
+    @Override
+    protected int engineGetBlockSize()
+    {
+        return BLKLEN;
+    }
+
+    @Override
+    protected int engineGetOutputSize(int inputLen)
+    {
+        return BLKLEN;
```

Comment:
I don't think this is right.  For CTR it should be equal to inputLen I think?

Reply:
Probably... I have to admit I haven't really read all the interfaces. For regular AES it's correct though as  the block size is always 128bit, not exactly sure what happens with GCM/AAED.

Reply:
This is "if I encrypt N `inputSize` bytes, how many bytes will I get out?"  So for CTR it's always equal to `inputSize`.

For GCM/AEAD it's more complicated, because of the tag.

Reply:
Ah, right.

Reply:
That said, as long as our code never calls `Cipher.getInputSize()` on the cipher this SPI implements, this method will never be called, so I don't think it matters much.

Reply:
Seems we currently don't, but I fixed it anyway.

Reply:
We'll have to revist if the same class should implement GCM via OpenSSL, but I think this would then be inputLen + blocksize.